src/chkstat.cpp

```diff
@@ -21,1365 +21,1248 @@
+                std::cerr << ctx.subpath << ": on an insecure path - " << path << " has different non-root owner who could tamper with the file." << std::endl;
+                return false;
             }
-            if (inpart)
-                pcnt++;
-            if (pcnt != 3)
+        }
+
+        if (ctx.status.isLink())
+        {
+            // If the path configured in the permissions configuration is a symlink, we don't follow it.
+            // This is to emulate legacy behaviour: old insecure versions of chkstat did a simple lstat(path) as 'protection' against malicious symlinks.
+            if (is_final_path_element || ++link_count >= 256)
+                // TODO: would an excess link count not warrant a complaint?
+                return false;
+
+            // Don't follow symlinks owned by regular users.
+            // In theory, we could also trust symlinks where the owner of the target matches the owner
+            // of the link, but we're going the simple route for now.
+            if (ctx.status.st_uid && ctx.status.st_uid != m_euid)
             {
-                BAD_LINE();
-                continue;
+                const auto path = getPathFromProc(pathfd);
+                std::cerr << ctx.subpath << ": on an insecure path - " << path << " has different non-root owner who could tamper with the file." << std::endl;
+                return false;
             }
-            part[3] = part[2];
-            part[2] = strchr(part[1], ':');
-            if (!part[2])
-                part[2] = strchr(part[1], '.');
-            if (!part[2])
+
+            std::string link(PATH_MAX, '\0');
+            const auto len = ::readlinkat(pathfd.get(), "", &link[0], link.size());
+
+            if (len <= 0 || static_cast<size_t>(len) >= link.size())
+                // TODO: would an error read the link not warrant a complaint?
+                return false;
+
+            link.resize(static_cast<size_t>(len));
+            stripTrailingSlashes(link);
+
+            if (link[0] == '/')
             {
-                BAD_LINE();
-                continue;
+                // absolute link, need to continue from a new root
+                pathfd.close();
             }
-            *part[2]++ = 0;
-            mode = (mode_t)strtoul(part[3], part + 3, 8);
-            if (mode > 07777 || part[3][0])
+            else
             {
-                BAD_LINE();
-                continue;
+                // relative link: continue relative to the parent directory
+
+                // we encountered a link directly below /, simply continue
+                // from the root again
+                if (parentfd.invalid())
+                {
+                    pathfd.close();
+                }
+                else
+                {
+                    pathfd.set( dup(parentfd.get()) );
+                }
+
+                // make the link absolute to fulfill the expectations of the
```

Comment:
the resulting path isn't actually absolute - it's relative to `parentf`... it just has to start with a slash because that's how the loop peels off leading path components.

Reply:
I understood that already. Let's rephrase that to `make the path absolute` to avoid any confusion.

Reply:
We may have different definitions of an absolute path:

- a path starting with a slash

vs.
- a path that needs to be resolved from the file system root

Adding a slash in front makes the path absolute according to the first definition, but not at all according to the second one.

Please just don't use the word 'absolute' to avoid confusion.

Reply:
I'm looking at this purely from an "addressing scheme" point of view. If such a path exists or not or is used as an absolute path or not is another question. The notion of `path_rest` and how it is used in the function make it obvious to me what is meant with it.

I'm not attached to the word absolute so let's change it.